SI-7275 allow flattening of blocks with ..$

densh opened the pull request:
This commit extends current splicing rules to allow flattening of
trees into other trees.

Without such support it is impossible to correctly use vals with
patterns and use it in other location as they could expand into
multiple-statement blocks:

``` scala
    scala> q"val (a, b) = (1, 2)"
    res0: reflect.runtime.universe.Tree =
    {
      <synthetic> <artifact> private[this] val x$1 = scala.Tuple2(1, 2):
    @scala.unchecked match {
        case scala.Tuple2((a @ _), (b @ _)) => scala.Tuple2(a, b)
      };
      val a = x$1._1;
      val b = x$1._2;
      ()
    }

    scala> q"..$res0; println(a + b)"
    res1: reflect.runtime.universe.Tree =
    {
      <synthetic> <artifact> private[this] val x$1 = scala.Tuple2(1, 2):
    @scala.unchecked match {
        case scala.Tuple2((a @ _), (b @ _)) => scala.Tuple2(a, b)
      };
      val a = x$1._1;
      val b = x$1._2;
      println(a.$plus(b))
    }
```

review by @xeno-by

densh committed:
Tag synthetic unit with attachment

This makes it easy to differentiate unit inserted by a compiler vs unit
written by the user. Useful for quasiquotes and pretty printing.

Additionally SyntacticBlock extractor is changed to treat EmptyTree as
zero-element block.

densh committed:
Refactor reification of high-cardinality holes

New approach makes iterated function much more clear through aggressive
code reuse, recursion and large descriptive comment on top of it.

densh committed:
SI-7275 allow flattening of blocks with ..$

This commit extends current splicing rules to allow flattening of
trees into other trees.

Without such support it is impossible to correctly create vals with
patterns and use it in other location as they could expand into
multiple-statement blocks:

    scala> q"val (a, b) = (1, 2)"
    res0: reflect.runtime.universe.Tree =
    {
      <synthetic> <artifact> private[this] val x$1 = scala.Tuple2(1, 2):
    @scala.unchecked match {
        case scala.Tuple2((a @ _), (b @ _)) => scala.Tuple2(a, b)
      };
      val a = x$1._1;
      val b = x$1._2;
      ()
    }

    scala> q"..$res0; println(a + b)"
    res1: reflect.runtime.universe.Tree =
    {
      <synthetic> <artifact> private[this] val x$1 = scala.Tuple2(1, 2):
    @scala.unchecked match {
        case scala.Tuple2((a @ _), (b @ _)) => scala.Tuple2(a, b)
      };
      val a = x$1._1;
      val b = x$1._2;
      println(a.$plus(b))
    }

xeno-by commented on src/compiler/scala/tools/reflect/quasiquotes/Reifiers.scala:
```diff
@@ -189,8 +189,8 @@ trait Reifiers { self: Quasiquotes =>
         reifyTree(other)
-      case Block(stats, last) =>
-        reifyBuildCall(nme.SyntacticBlock, stats :+ last)
+      case SyntacticBlock(stats) if stats.size > 1 =>
```
Iirc Paul or someone else has told me that length is always preferable to size, but I no longer remember the reason. @retronym, maybe you have an idea why it might be the case?

xeno-by commented on src/reflect/scala/reflect/internal/StdAttachments.scala:
```diff
@@ -36,6 +36,10 @@ trait StdAttachments {
    */
   case object ForAttachment extends PlainAttachment
 
+  /** Identifies unit constants which were inserted by the compiler.
```
I think this comment would benefit from a list of situations when scalac inserts synthetic units.

xeno-by commented on src/compiler/scala/tools/reflect/quasiquotes/Holes.scala:
```diff
@@ -118,21 +119,64 @@ trait Holes { self: Quasiquotes =>
+      case (DotDot, tpe @ IterableType(LiftedType(lift))) => mapF(toList(tree, tpe), lift)
+      case (DotDot, LiftedType(lift))                     => toStats(lift(tree))
+      case (DotDotDot, tpe @ IterableType(inner))         => mapF(toList(tree, tpe), t => iterated(DotDot, t, inner))
+      case (DotDotDot, LiftedType(lift))                  => mapF(toStats(lift(tree)), toStats)
```
Very impressive code reduction!

xeno-by commented:
I think this is an important improvement to usability of quasiquotes. LGTM!

densh committed:
Address pull request feedback

densh commented:
PLS REBUILD/pr-scala@ef118b8

scala-jenkins commented:
(kitty-note-to-self: ignore 33129167)
:cat: Roger! Rebuilding pr-scala for ef118b8c. :rotating_light:

xeno-by commented:
LGTM

densh committed:
Reify all blocks as syntactic blocks

Previous version of the pattern wasn't precise enough due to the
fact that synthetic units are now erased by syntactic block and
this could cause incorrect reification for cases like:

   val x = { val y = 1 }

Here syntactic block would extract one element but we still need to
reify it through syntactic block endpoint. So we can't filter based
on the number of elements extracted but rather filter on type of a
tree.

densh commented:
@xeno-by Have just found a minor bug in the new reification of the blocks and hotfixed it by 3074857.

densh commented:
PLS REBUILD/pr-scala@0a529b9

scala-jenkins commented:
(kitty-note-to-self: ignore 33214257)
:cat: Roger! Rebuilding pr-scala for 0a529b90. :rotating_light:

densh committed:
Addresses feedback from Jason

1. Adds tests for new synthetic unit stripping.
2. Marks implementation-specific parts of Holes as private.
3. Trims description of iterated method a bit.
4. Provides a bit more clear wrapper for q interpolator.
5. Refactors SyntacticBlock, adds documentation.
6. Makes q"{ ..$Nil }" return q"" to be consist with extractor.

densh commented:
@retronym just pushed a commit that address your feedback.
